### PR TITLE
feat(cfw): add new resource firewall

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/cfw' TESTARGS='-run TestAccFirewall_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run TestAccFirewall_basic -timeout 360m -parallel 4
=== RUN   TestAccFirewall_basic
=== PAUSE TestAccFirewall_basic
=== CONT  TestAccFirewall_basic
--- PASS: TestAccFirewall_basic (385.23s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       385.274s

make testacc TEST='./huaweicloud/services/acceptance/cfw' TESTARGS='-run TestAccFirewall_prePaid'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run TestAccFirewall_prePaid -timeout 360m -parallel 4
=== RUN   TestAccFirewall_prePaid
=== PAUSE TestAccFirewall_prePaid
=== CONT  TestAccFirewall_prePaid
--- PASS: TestAccFirewall_prePaid (469.03s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       469.132s
```

### CODE REVIEW

File: docs/resources/cfw_firewall.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+* `east_west_firewall_inspection_vpc_id` - The east-west firewall inspection VPC ID.
+
+<a name="Firewall_ProtectObject"></a>
+The `protect_objects` block supports:
+
+* `object_id` - The protected object ID.
+
+* `object_name` - The protected object name.
+
+* `type` - The Project type.
```
Comment (deer-hang) [inline docs/resources/cfw_firewall.md]:
```suggestion
* `type` - The project type.

```

File: docs/resources/cfw_firewall.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
+API response, security or some other reason. The missing attributes include:
+`flavor`, `tags`, `period_unit`, `period` and `auto_renew`. It is generally
+recommended running `terraform plan` after importing an CFW firewall. You can then decide if changes should be applied to
+the firewall, or the resource definition should be updated to align with the firewall. Also you can ignore changes as
+below.
+
+```
+resource "huaweicloud_cce_cluster" "cluster_1" {
```
Comment (deer-hang) [inline docs/resources/cfw_firewall.md]:
```suggestion
resource "huaweicloud_cfw_firewall" "test" {

```

The following `ignore_changes` should also be changed.

File: huaweicloud/services/acceptance/acceptance.go [hunk 3]
```diff
@@ -764,6 +765,13 @@ func TestAccPreCheckCfw(t *testing.T) {
 	}
 }
 
+// lintignore:AT003
+func TestAccPreCheckCfwEastWestFirewall(t *testing.T) {
+	if HW_CFW_EAST_WEST_FIREWALL == "" {
+		t.Skip("HW_CFW_INSTANCE_ID must be set for CFW east-west firewall acceptance tests")
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/acceptance.go]:
```suggestion
		t.Skip("HW_CFW_EAST_WEST_FIREWALL must be set for CFW east-west firewall acceptance tests")
```

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+	getFirewallPath := getFirewallClient.Endpoint + getFirewallHttpUrl
+	getFirewallPath = strings.ReplaceAll(getFirewallPath, "{project_id}", getFirewallClient.ProjectID)
+	getFirewallPath += fmt.Sprintf("?offset=0&limit=10&service_type=0&fw_instance_id=%s", state.Primary.ID)
+
+	getFirewallOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go]: Better to remove `OkCodes` when success code is 200. Change the other parts.

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	}
+	getFirewallsResp, err := getFirewallClient.Request("GET", getFirewallPath, &getFirewallOpt)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving Firewalls: %s", err)
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go]:
```suggestion
		return nil, fmt.Errorf("error retrieving firewalls: %s", err)
```

Please check the other logging message.

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+  tags = {
+    key = "value"
+    foo = "bar"
+  }
+
+  charging_mode = "prePaid"
+  period_unit = "month"
+  period = 1
+  auto_renew = false
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go]: Parameter alignment.

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go [hunk 7]
```diff
@@ -0,0 +1,847 @@
...
+		return diag.FromErr(err)
+	}
+
+	orderID := utils.PathSearch("order_id", createFirewallRespBody, "")
+	jobID := utils.PathSearch("job_id", createFirewallRespBody, "")
+
+	bssClient, err := cfg.BssV2Client(cfg.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating BSS v2 client: %s", err)
+	}
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go]: Better to remove line 229-232 to line 235

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go [hunk 8]
```diff
@@ -0,0 +1,847 @@
...
+		createEastWestFirewallOpt.JSONBody = utils.RemoveNil(buildCreateEastWestFirewallBodyParams(d))
+		createEastWestFirewallResp, err := createFirewallClient.Request("POST", createEastWestFirewallPath, &createEastWestFirewallOpt)
+		if err != nil {
+			return diag.Errorf("error creating east-west Firewall: %s", err)
+		}
+
+		_, err = utils.FlattenResponse(createEastWestFirewallResp)
+		if err != nil {
+			return diag.FromErr(err)
+		}
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go:268]: It seems that there is no need to do this verification.

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go [hunk 9]
```diff
@@ -0,0 +1,847 @@
...
+	return resourceFirewallUpdate(ctx, d, meta)
+}
+
+func buildCreateFirewallBodyParams(d *schema.ResourceData, cfg *config.Config) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"name":                  d.Get("name"),
+		"flavor":                buildCreateFirewallRequestBodyFlavor(d.Get("flavor")),
+		"tags":                  utils.ExpandResourceTags(d.Get("tags").(map[string]interface{})),
+		"charge_info":           buildCreateFirewallRequestBodyChargeInfo(d),
+		"enterprise_project_id": utils.ValueIngoreEmpty(common.GetEnterpriseProjectID(d, cfg)),
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go]:
```suggestion
		"enterprise_project_id": utils.ValueIngoreEmpty(cfg.GetEnterpriseProjectID(d)),
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go [hunk 10]
```diff
@@ -0,0 +1,847 @@
...
+
+			pendingStatus := []string{
+				"Running",
+			}
+			if utils.StrSliceContains(pendingStatus, status) {
+				return createFirewallWaitingRespBody, "PENDING", nil
+			}
+
+			return createFirewallWaitingRespBody, status, nil
+		},
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go:412]: I think this method can be optimized.
- The `client` can be passed in through parameters, no need to create a new one every time.
- Function `StrSliceContains` seems unnecessary.

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go [hunk 11]
```diff
@@ -0,0 +1,847 @@
...
+
+	"github.com/chnsz/golangsdk"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+const (
+	FirewallNotExistsCode = "CFW.00200005"
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go:30]:
```suggestion
	firewallNotExistsCode = "CFW.00200005"
```

File: docs/resources/cfw_firewall.md (hunk 1 above)
Comment (Jason-Zhang9309) [reply docs/resources/cfw_firewall.md]: fixed

File: docs/resources/cfw_firewall.md (hunk 2 above)
Comment (Jason-Zhang9309) [reply docs/resources/cfw_firewall.md]: fixed

File: huaweicloud/services/acceptance/acceptance.go (hunk 3 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/acceptance.go]: fixed

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go (hunk 4 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go]: fixed

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go (hunk 5 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go]: fixed

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_firewall_test.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go (hunk 7 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go]: sure

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go (hunk 9 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go (hunk 8 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go:268]: fixed

**Pull request merged**